pkg/controller/compute-machineset/compute_machineset_controller.go

```diff
@@ -0,0 +1,519 @@
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	// create/delete Machine objects
+	err := c.syncHandler(key.(string))
+	c.handleErr(err, key)
+
+	return true
+}
+
+func (c *Controller) ensureMachinesAreDeleted(ms *clusteroperator.MachineSet) error {
+	machines, err := controller.MachinesForMachineSet(ms, c.machineLister)
+	if err != nil {
+		return err
+	}
+	errs := []error{}
+	for _, machine := range machines {
+		if machine.DeletionTimestamp == nil {
+			err = c.client.ClusteroperatorV1alpha1().Machines(machine.Namespace).Delete(machine.Name, &metav1.DeleteOptions{})
+			if err != nil && !errors.IsNotFound(err) {
+				errs = append(errs, err)
+			}
+		}
+	}
+	if len(errs) > 0 {
+		return utilerrors.NewAggregate(errs)
+	}
+	return nil
+}
+func (c *Controller) syncMachineSet(key string) error {
+	c.logger.WithField("key", key).Debug("synching machineset")
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		c.logger.Errorf("error parsing namespace out of machineset name %v", key)
+		return err
+	}
+
+	ms, err := c.machineSetsLister.MachineSets(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		c.logger.WithField("key", key).Errorf("couldn't find machineset %v", key)
+		c.expectations.DeleteExpectations(key)
+		return err
+	}
+	if err != nil {
+		c.logger.Errorf("problem fetching machineset %v: %v", key, err)
+		return err
+	}
+
+	msLog := colog.WithMachineSet(c.logger, ms)
+
+	if ms.DeletionTimestamp != nil {
+		msLog.Debug("machineset has been deleted. Ensuring that its machines are also deleted.")
+		return c.ensureMachinesAreDeleted(ms)
+	}
+
+
+	// get all the machines in our current namespace
+	allMachines, err := c.machineLister.Machines(namespace).List(labels.Everything())
+	if err != nil {
+		c.logger.WithField("machineset", ms).Errorf("error while trying to get list of all machines")
+		return err
+	}
+
+	// get all machines which are children of this machineset
+	var filteredMachines []*clusteroperator.Machine
+	for _, machine := range allMachines {
+		if machine.DeletionTimestamp != nil {
+			continue
+		}
+		controllerRef := metav1.GetControllerOf(machine)
+		if controllerRef == nil {
+			c.logger.WithField("machineset", name).Debugf("no controller ref for machine %v", machine)
+			continue
+		}
+		if ms.UID != controllerRef.UID {
+			//c.logger.WithField("machineset", name).Debug("the kid is not my son %v", machine)
+			continue
+		}
+		filteredMachines = append(filteredMachines, machine)
+	}
+
+	var manageMachinesErr error
+
+	machineSetNeedsSync := c.expectations.SatisfiedExpectations(key)
+
+	if machineSetNeedsSync && ms.DeletionTimestamp == nil {
+		manageMachinesErr = c.manageMachines(filteredMachines, ms)
+	}
+
+	// update Status field
```

Comment:
Well want to figure out what to do about the accept controller being blocked on the provisioning of the machineset. We need to either (1) run the accept controller on machines instead of machinesets or (2) track in the individual machines whether they have been provisioned and carry that back to the machineset.